### ISSUE TITLE
Sharing an RDS instance between two workloads (or providing a read replica for one of them)

### ISSUE THREAD

**Issue by lexicalunit** [aws/copilot-cli#4273]
I've got a backend service that uses an RDS database, Aurora Serverless. I was able to set that up easily enough by going through the `copilot storage init` wizard. However I also have a load balanced web service that needs read-only access to that very same database. I tried running `copilot storage init` but I think it's just setting up a totally separate database. I thought maybe I could go into the manifest file for the web service and add a reference to the storage there somehow, but from the docs it looks like the storage section in the manifest file is only used for EFS? https://aws.github.io/copilot-cli/docs/developing/storage/

Is there a way to do this?

Maybe I just need to copy/paste `/copilot/backend-service-name/addons/cluster.yml` to `/copilot/web-app-name/addons/cluster.yml`? So they use the same "addon"? 🤔

**Comment by thesabbir** [comment]
You can use the same storage secret in multiple services.

Edit: You need to modify the addon for the load-balanced web service first to allows ingress from the EnvironmentSecurityGroup as described by @efekarakus

**Comment by efekarakus** [comment]
Hi @lexicalunit @thesabbir 

> Maybe I just need to copy/paste /copilot/backend-service-name/addons/cluster.yml to /copilot/web-app-name/addons/cluster.yml? So they use the same "addon"? 🤔

That will create a separate RDS cluster unfortunately so they won't share the same database 😅 

There are few different ways of achieving this functionality, I'll write the various options and tradeoffs:

1. **[preferred] The backend service can communicate with the load balanced web service using [service connect](https://aws.github.io/copilot-cli/docs/developing/svc-to-svc-communication/)**. So the RDS cluster addon would remain only accessible by the load balanced web service. Instead, the load balanced web service can provide an API that the backend service can use to read data from. For example, if your load balanced web service is called "spellbot" and exposes port 80 in `image.build`, then we can add a new API in the spellbot service for say "GET /wizards". Then the backend service can make a request to that path:
   ```go
   resp, err := http.Get("http://spellbot/wizards") 
   ```
<p align="center">
   <img src="https://user-images.githubusercontent.com/879348/207929379-62907835-cd59-4d8e-b186-bdc6ac985bb7.png" width="400px" />
</p>
   The tradeoff is that it's a bit more work upfront then just letting the backend service access the database.


2. An alternative is what my teammate @Lou1415926 proposes here: https://github.com/aws/copilot-cli/issues/2495#issuecomment-866188721 - enabling ingress from the EnvironmentSecurity Group. 
   We can modify the addon for the load balanced web service such that the security group for the RDS addon allows ingress from the `EnvironmentSecurityGroup`. This way all services that get deployed to the same environment will have access to the database. And like @thesabbir wrote, we can inject the same secret to the backend service's manifest:
   ```yaml
   secrets:
      DB: <secret ARN> # For alternative ways of injecting secrets: https://aws.github.io/copilot-cli/docs/developing/secrets/#in-secrets-manager
   ```
   The tradeoff is that now both microservices can read/write from the database so we created tighter coupling between services.
   
   
Hope this helps!

**Comment by thesabbir** [comment]
@efekarakus Thank you so much. Waiting for the environment add-ons to land! 
:)

**Comment by efekarakus** [comment]
I moved the answer to the show-and-tell section as well to make it easier to discover: https://github.com/aws/copilot-cli/discussions/4275

**Comment by lexicalunit** [comment]
> The backend service can communicate with the load balanced web service using [service connect](https://aws.github.io/copilot-cli/docs/developing/svc-to-svc-communication/).

I would use this for most normal services but it doesn't work for my particular application. Here's what I have:

- A backend service that is a Discord bot and it also uses a database. Users interact with it through discord. It writes and reads to/from the database. It has some minimal facility to make network requests outside of Discord websockets, but not much. Generally it's not set up to accept HTTP requests nor send them.
- A front end web API that essentially provides a view into the data that the bot has created. You hit a page, you get some static content generated from what's in the database. Read only. Doesn't communicate with Discord bot at all.

So the source of truth is the database and the primary actor on that database is the backend service. If I wanted to do something similar to what you have in your diagram, the arrow between services would be pointed the other way, and the database would be hanging off the backend service. And to do that I would have to build out a whole web API into the discord bot, but it's already over-taxed handing Discord requests. So that's not ideal.

**Comment by Lou1415926** [comment]
You can now use environment addons (released in [v1.25.0](https://github.com/aws/copilot-cli/releases/tag/v1.25.0)) to achieve this goal! You can also use `storage init` for a guided experience to create a storage resource that is managed on the environment level (released in [v1.26.0](https://github.com/aws/copilot-cli/releases/tag/v1.26.0)). See also our [blog posts](https://aws.github.io/copilot-cli/blogs/release-v126/) to learn more ❤️
